File: 210.py
# time: V + E
# space: V + E
from collections import deque
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = {i:[] for i in range(numCourses)} 
        indegree = {i:0 for i in range(numCourses)} 
        for prerequisite in prerequisites:
            graph[prerequisite[1]].append(prerequisite[0])
            indegree[prerequisite[0]] += 1
        
        queue = deque([node for node in indegree if indegree[node] == 0])
        topo_loop = []

        while queue: 
            node = queue.popleft()
            topo_loop.append(node)
            for neighbour in graph[node]: 
                indegree[neighbour] -= 1
                if indegree[neighbour] == 0:
                    queue.append(neighbour)
        
        if len(topo_loop) == numCourses:
            return topo_loop
        else:
            return []

File: 210.cpp
// time: V + E
// space: V + E
// Khan做法, 很容易背板的
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        //graph 
        vector<vector<int>> g(numCourses);
        for(auto & p : prerequisites) {
            int a = p[0], b = p[1];
            g[b].push_back(a);
        }

        vector<int> indegree(numCourses), order;
        for (int i = 0; i < numCourses; i++) {
            for (int v : g[i]) {
                indegree[v]++;
            }
        }

        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                q.push(i);
            }
        }

        while(!q.empty()) {
            int node = q.front(); q.pop();
            order.push_back(node);
            for (int neighbor : g[node]) {
                indegree[neighbor]--;
                if(indegree[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }

        return (order.size() == numCourses) ? order : vector<int>();
    }
};